Fix StreamPriorityKey::cmp implementation

PSUdaemon opened the pull request:
Fix `StreamPriorityKey::cmp` implementation so it doesn't violate Rust's `Ord`
trait contract while keeping the intended round-robin behavior.

This is based on the discussion that was had in #2282. @antoniovicente had some
concerns about the implementation with respect to the `Ord` trait contract.

This is my attempt at fixing that and maintaining the round-robin requirement.

- Ordering comes from a derived `Ord` on a new position type, so the contract
  holds without a hand-written comparison.
- The sequence field is per queue, so reading from a stream doesn't reorder what
  we send on it.
- `cycle_flushable()` became `cycle_priority()`. It takes the queue that was
  serviced, and runs every time we service a stream rather than only when the
  stream stays queued.
- `stream_readable_next()` and `stream_writable_next()` take a stream off the
  queue as they hand it out. The old reinsert put it back by accident, and only
  for incremental streams, so membership follows eligibility now instead.
- Moved `stream_priority()` into `StreamMap::set_priority()`, so the priority
  tests don't need their own copy of it.

The h3 change is a test, see the comment on the thread.

The branch is rebased onto master and squashed, so the existing review threads
point at lines that no longer exist. Happy to re-open any of them.

LPardue commented on quiche/src/h3/mod.rs:
```diff
@@ -3825,33 +3825,30 @@ mod tests {
             more_frames: true,
         };
         assert_eq!(ev, ev_headers);
+        assert_eq!(s.poll_server(), Ok((0, Event::Data)));
```
Needing to change this test seems a bit fishy to me. 

While RFC 9218 primarily targets prioritization of serving responses, we apply stream scheduling to all QUIC streams. The default priority for a new stream is, in quiche's model, for urgency=127 and incremental=true (see https://github.com/cloudflare/quiche/blob/86a79a7a049701baf608d6e88a917fa96fe804b6/quiche/src/stream/mod.rs#L798). That should trigger round-robin sending of all request data.

In the previous test, we saw the server reading events in the round-robin order of stream sends. Although that's only partly true because the recv_body() function attempts to greedily read all available data frames independent of the peer's sender scheduling. 

The changes to this test seem to imply that sending has turned into FIFO, where HEADERS and DATA are sent together (or, the peer is somehow receiving them together). It's not abundantly clear to me what's going on. I think it would pay to get some more understanding why this test triggered some change while others didn't

PSUdaemon replied:
> A stop gap might be, on the reader side, to cycle the stream after generating an Event::Headers?

Yeah, that's an interesting idea. Are you looking to have that added in this PR, or as a followup one?

LPardue replied:
I think for parity with the current behaviour, we'd need that in this PR. Would you mind trying it and see if it achieves the old ordering?

PSUdaemon replied:
The headers first behavior you asked for is now baked into the design, so I removed the explicit code in the h3 module. I replaced it with a test to make sure we don't accidentally lose it again.

LPardue commented on quiche/src/tests.rs:
```diff
@@ -7804,9 +7804,10 @@ fn last_tx_data_larger_than_tx_data(
+    // order, which may vary; we just need some data to be sent.
     let (len, _) = pipe.server.send(&mut buf).unwrap();
-    assert_eq!(len, 1200);
+    assert!(len > 0);
```
This also seems suspect - the round robin should be determininstic no?

PSUdaemon committed:
Key the stream priority queues on a total order.

StreamPriorityKey::cmp returned Greater for any two same-urgency
incremental streams, so a and b each compared greater than the other, and
across three such streams the relation was not transitive either. The
implementation was deliberate: removing a stream and reinserting it under
that comparison landed it at the back of its priority group, which is how
round-robin scheduling was implemented. Keying a red-black tree on a
comparison that is not a total order leaves the order it produces
undefined, so the scheduling that depended on it was not guaranteed.

Ordering now comes from StreamPriorityPosition, and the three tree
adapters return one instead of cloning the key, which also takes the
clone off every comparison in a tree descent. Its field order states the
comparison: urgency, then non-incremental ahead of incremental, then a
round-robin sequence, then stream ID. Deriving Ord over those fields
makes the contract hold by construction, and StreamPriorityKey no longer
implements Ord, PartialOrd, Eq or PartialEq at all, so comparing two keys
is a compile error rather than a question. Round-robin becomes explicit
rather than a side effect of reinsertion: a stream takes a sequence
number from the connection when it is created and a fresh one each time
it is serviced, which places it behind the streams it was ahead of.
Non-incremental streams have no round-robin position, so the sequence is
masked out for them, keeping them ordered by stream ID.

Two things the reinsertion trick did for free have to be arranged once
the position is explicit.

A stream was only shuffled while it stayed in its queue, which was enough
when every insert went to the back. Against a position that persists, a
stream that fully drained, or that ran out of flow control capacity,
keeps the position it held on the way out and sorts ahead of streams that
have been waiting since before it was served.

So a stream is cycled whenever it is serviced, and cycle_priority decides
the rest: whether the service left the stream in the queue, whether the
stream still exists, and whether it is incremental at all, since a
non-incremental stream has no round-robin position to move. A service
point is one unconditional statement, which leaves omitting one as the
only way to get it wrong. Each way of getting a per-site condition wrong
had been a bug, and the conditions were also where the sites disagreed
with each other, two reading Stream::incremental and one the key's copy.
There are four service points: after emitting a stream's data, after
reading from it, after writing to it, and when stream_writable_next hands
it to the application, which takes it out of the writable queue as its
form of service.

Membership has to follow eligibility, which the trick also supplied by
accident. Its reinsert put back any stream that had left the queue,
including one that left for a reason other than running out of work:
stream_readable_next and stream_writable_next take a stream out as they
hand it to the application. Cycling cannot do that job, since cycling a
stream into a queue it is not eligible for would be worse. So the recv
and send paths now insert the stream when it is still eligible and remove
it when it is not, rather than only removing. Without that, a partial
read after stream_readable_next left the stream out of the readable queue
with data still buffered, invisible to readable() as well, which offers
no re-arm contract at all. On the send path this generalises a special
case that was already there for streams held out by connection-level
capacity. The trick only re-armed incremental streams, since that is what
it shuffled, so this also stops a non-incremental stream being stranded
the same way.

Streams that have not yet been serviced in a queue are now ordered by
when they were created rather than by when they first entered it. Under
the trick those coincided, since first entry was when the insert
happened; an explicit position is stamped at creation instead. Both are
orderings of streams that have waited equally, RFC 9218 says nothing
about the order within an incremental group, and round-robin is
unaffected once each has been served once, but a packet capture shows it.

The three queues also have to hold separate positions, since reinserting
into one tree said nothing about the others. Sharing one sequence across
them would mean servicing a stream in one queue moved it in the other
two, so that reading from a stream reordered what was sent on it. Send
scheduling should follow the peer's priority signals, not how promptly
the local application drains its receive buffers, so each queue gets its
own sequence field.

Cycling one queue therefore has to leave the other two alone, and should
not cost more than the remove and reinsert it replaces. Because each
queue reads only its own field, cycle_priority can unlink the stream from
that one queue, store the new sequence through the Arc, and relink it:
the same two tree operations as before, with the key untouched and the
Arc held by each queue still the one the stream holds. Rebuilding the key
instead would take the stream out of all three trees, allocate, and
insert it into each of them again. The fields are AtomicU64 because the
store goes through a shared Arc and because the key has to stay Sync for
Connection to be Send. Storing under a linked node would leave that queue
sorted by a value the node no longer has, so set_sequence asserts the
node is unlinked. Setting a stream's priority still rebuilds the key,
since urgency and the incremental flag apply to every queue; that path
moves from Connection::stream_priority into StreamMap::set_priority,
alongside the rest of the queue bookkeeping.

The priority tests in stream/mod.rs held hand-rolled copies of
stream_priority, in order to reach a stream and the queues it is in
together. They call set_priority now. What they assert is unchanged: all
28 expected orderings are the orderings they expected before, and the
tests that pin the scheduling behaviour pass against master as well.

The PTO probe in last_tx_data_larger_than_tx_data now carries stream 8
rather than stream 4. Stream 4 was cycled to the back of the flushable
queue on each packet that carried its 10000 bytes, while stream 8 has not
been serviced at all, so round-robin picks stream 8, which holds 800
bytes against what is left of the connection's flow control credit. Only
one STREAM frame goes in a packet, so the probe is those 800 bytes plus
48 of framing: a 5 byte STREAM header, 9 bytes of ACK and DATA_BLOCKED,
an 18 byte short header, and a 16 byte tag. The 1200 asserted before was
the size of the buffer passed to send() rather than a length the stream
chose: stream 4's 1200 buffered bytes wanted 1248 with the same framing,
and filled the buffer instead.

The HTTP/3 change is one test. A server polling concurrent requests sees
the headers of every request before the body of any of them, which lets
it route or reject on headers alone; that ordering comes from the
transport cycling incremental streams as poll() reads them, and losing it
is not obvious from either layer alone.
